pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -466,12 +467,22 @@ public void openCursorFailed(ManagedLedgerException exception, Object ctx) {
     private CompletableFuture<? extends Subscription> getNonDurableSubscription(String subscriptionName, MessageId startMessageId) {
         CompletableFuture<Subscription> subscriptionFuture = new CompletableFuture<>();
 
+        // Create a new non-durable cursor only for the first consumer that connects
         Subscription subscription = subscriptions.computeIfAbsent(subscriptionName, name -> {
-            // Create a new non-durable cursor only for the first consumer that connects
             MessageIdImpl msgId = startMessageId != null ? (MessageIdImpl) startMessageId
                     : (MessageIdImpl) MessageId.latest;
 
-            Position startPosition = new PositionImpl(msgId.getLedgerId(), msgId.getEntryId());
+            long ledgerId = msgId.getLedgerId();
+            long entryId = msgId.getEntryId();
+            if (msgId instanceof BatchMessageIdImpl) {
```

Comment:
I think broker is batch-msg agnostic as it treats it as a normal msg (client manages batch-msg business) and it would be better to avoid introducing `BatchMessageIdImpl` at broker-side. Can't we always assign `entryId= entryId-1` while creating  `NonDurableSubscription` ?

Reply:
> Can't we always assign entryId= entryId-1 while creating NonDurableSubscription? 

The problem is that you can have these 2 scenarios: 
 * Without batching : `startMessageId = (ledgerId, entryId)`. eg. `(3, 4)`
 * With batching : `startMessageId = (ledgerId, entryId, batchIndex)`. eg. `(3,4,5)`

In the first case, it will mean that `(3,4)` was already consumed. So next message to be sent will be `(3,5)`.

In the case of batching, `(3,4,5)` is the start message. With the old logic, the broker would start delivering at `(3,5,0)` while there might still be messages to read at `(3,4,6)`..

Reply:
> Without batching : startMessageId = (ledgerId, entryId). eg. (3, 4)

yes, I understand that batch-msg might not be consumed fully at client so, we need to go 1 msg back. 
If we do the same for non-batch then `(3,4)` will be delivered again which will be duplicate. Is there an issue if broker delivers duplicate msg? because anyway for durable cursor, broker can deliver duplicate msg by reseting cursor-position at `marDeletePosition`. So, duplicate msg is the only conern to set `entryId=entryId-1`?

Reply:
> Without batching : startMessageId = (ledgerId, entryId). eg. (3, 4)

yes, I understand that batch-msg might not be consumed fully at client so, we need to go 1 msg back. 
If we do the same for non-batch then `(3,4)` will be delivered again which will be duplicate. Is there an issue if broker delivers duplicate msg? because anyway for durable cursor, broker can deliver duplicate msg by reseting cursor-position at `marDeletePosition`. So, duplicate msg is the only conern to set `entryId=entryId-1`?

Reply:
Well, the purpose of reader is to position exactly on the correct message :) That enables to avoid dups in the consuming side.

Reply:
In any case, that is an issue even with the current Java reader. I have another PR to fix that as well.